src/Expression/StructuredExpression.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Db\Expression;
+
+use Traversable;
+use Yiisoft\Db\Schema\Column\ColumnSchemaInterface;
+
+use function array_key_exists;
+use function array_keys;
+use function is_iterable;
+use function iterator_to_array;
+
+/**
+ * Represents a structured type SQL expression.
+ *
+ * @see https://en.wikipedia.org/wiki/Structured_type
+ *
+ * For example:
+ *
+ * ```php
+ * new StructuredExpression(['price' => 10, 'currency_code' => 'USD']);
+ * ```
+ *
+ * Will be encoded to `ROW(10, USD)` in PostgreSQL.
+ */
+final class StructuredExpression implements ExpressionInterface
+{
+    /**
+     * @param ColumnSchemaInterface[] $columns
+     * @psalm-param array<string, ColumnSchemaInterface> $columns
+     */
+    public function __construct(
+        private mixed $value,
+        private string|null $type = null,
+        private array $columns = [],
+    ) {
+    }
+
+    /**
+     * The structured type name.
+     *
+     * Defaults to `null` which means the type is not explicitly specified.
+     *
+     * Note that in the case where a type is not specified explicitly and DBMS cannot guess it from the context,
+     * SQL error will be raised.
+     */
+    public function getType(): string|null
+    {
+        return $this->type;
+    }
+
+    /**
+     * The structured type columns that are used for value normalization and type casting.
+     *
+     * @return ColumnSchemaInterface[]
+     */
+    public function getColumns(): array
+    {
+        return $this->columns;
+    }
+
+    /**
+     * The content of the structured type. It can be represented as an associative array of structured type column names
```

Comment:
How else can it be represented? Why `mixed`?

Reply:
Replaced `mixed` to `array|object` and added description about possible values